shared/subspace-kzg/src/lib.rs

```diff
@@ -109,6 +104,229 @@ impl Polynomial {
+    }
+
+    /// Convenient conversion from slice of optional underlying representation to scalar for efficiency
+    /// purposes.
+    pub fn slice_option_from_repr(value: &[Option<FsFr>]) -> &[Option<Self>] {
+        // SAFETY: `Scalar` is `#[repr(transparent)]` and guaranteed to have the same memory layout
+        unsafe { mem::transmute(value) }
+    }
+
+    /// Convenient conversion from mutable slice of scalar to underlying representation for
+    /// efficiency purposes.
+    #[inline]
+    pub fn slice_mut_to_repr(value: &mut [Self]) -> &mut [FsFr] {
+        // SAFETY: `Scalar` is `#[repr(transparent)]` and guaranteed to have the same memory layout
+        unsafe { mem::transmute(value) }
+    }
+
+    /// Convenient conversion from mutable slice of underlying representation to scalar for
+    /// efficiency purposes.
+    #[inline]
+    pub fn slice_mut_from_repr(value: &mut [FsFr]) -> &mut [Self] {
+        // SAFETY: `Scalar` is `#[repr(transparent)]` and guaranteed to have the same memory layout
+        unsafe { mem::transmute(value) }
+    }
+
+    /// Convenient conversion from optional mutable slice of scalar to underlying representation for
+    /// efficiency purposes.
+    pub fn slice_option_mut_to_repr(value: &mut [Option<Self>]) -> &mut [Option<FsFr>] {
+        // SAFETY: `Scalar` is `#[repr(transparent)]` and guaranteed to have the same memory layout
+        unsafe { mem::transmute(value) }
+    }
+
+    /// Convenient conversion from optional mutable slice of underlying representation to scalar for
+    /// efficiency purposes.
+    pub fn slice_option_mut_from_repr(value: &mut [Option<FsFr>]) -> &mut [Option<Self>] {
+        // SAFETY: `Scalar` is `#[repr(transparent)]` and guaranteed to have the same memory layout
+        unsafe { mem::transmute(value) }
+    }
+
+    /// Convenient conversion from vector of scalar to underlying representation for efficiency
+    /// purposes.
+    pub fn vec_to_repr(value: Vec<Self>) -> Vec<FsFr> {
+        // SAFETY: `Scalar` is `#[repr(transparent)]` and guaranteed to have the same memory
+        //  layout, original vector is not dropped
+        unsafe {
+            let mut value = mem::ManuallyDrop::new(value);
+            Vec::from_raw_parts(
+                value.as_mut_ptr() as *mut FsFr,
+                value.len(),
+                value.capacity(),
+            )
+        }
+    }
+
+    /// Convenient conversion from vector of underlying representation to scalar for efficiency
+    /// purposes.
+    pub fn vec_from_repr(value: Vec<FsFr>) -> Vec<Self> {
+        // SAFETY: `Scalar` is `#[repr(transparent)]` and guaranteed to have the same memory
+        //  layout, original vector is not dropped
+        unsafe {
+            let mut value = mem::ManuallyDrop::new(value);
+            Vec::from_raw_parts(
+                value.as_mut_ptr() as *mut Self,
+                value.len(),
+                value.capacity(),
+            )
+        }
+    }
```

Comment:
> Vector storing transparent struct is the same as vector of the internal contents of that struct.

Yes, this is true, as long as we use `from_raw_parts()` (because field order isn’t guaranteed), and the arguments are taken unmodified from another vector (or the item layouts are compatible).

Here’s the full list of invariants [from the documentation for `from_raw_parts()`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts):

Most of them are ensured by re-using the unmodified fields from another vector, which is why I suggested adding “allocating via original vector ensures other invariants”:
- `ptr` must have been allocated using the global allocator, such as via the [alloc::alloc](https://doc.rust-lang.org/std/alloc/fn.alloc.html) function.
- `length` needs to be less than or equal to `capacity`.
- `capacity` needs to be the capacity that the pointer was allocated with.
- The allocated size in bytes must be no larger than `isize::MAX`. See the safety documentation of [pointer::offset](https://doc.rust-lang.org/std/primitive.pointer.html#method.offset).

And the rest are ensured if the layouts and valid values of the two item types are the same:
- `T` needs to have the same alignment as what `ptr` was allocated with. (`T` having a less strict alignment is not sufficient, the alignment really needs to be equal to satisfy the [dealloc](https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#tymethod.dealloc) requirement that memory must be allocated and deallocated with the same layout.)
- The size of `T` times the `capacity` (ie. the allocated size in bytes) needs to be the same size as the pointer was allocated with. (Because similar to alignment, [dealloc](https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#tymethod.dealloc) must be called with the same layout size.)
- The first `length` values must be properly initialized values of type `T`.

Reply:
I feel like those are automatically implied by `#[repr(transparent)]`. We have a simple struct with that is `#[repr(transparent)]` with `#[repr(C)]` struct inside of it. Might be just me, but "allocating via original vector ensures other invariants" doesn't really provide any additional context to me. We clearly just re-construct vector from the same exact raw parts. If the contents is the same, then alignment and other things are obviously also the same by extension.

Reply:
Sure, and my suggested wording is intended as a warning not to make any changes to any of the arguments to the unsafe function. For example, we can’t change the length or capacity, or copy the ptr in memory, because that could break their size, alignment, or allocator invariants.

As I said, this was a nitpick, so I’m not going to insist on any changes here.